### PR TITLE
sparse_retain op

### PR DESCRIPTION
**Pull request by reminisce**
Implemented sparse_retain op for row sparse tensors.

### CODE REVIEW

File: src/operator/tensor/indexing_op.cc [hunk 1]
```diff
@@ -264,5 +264,33 @@ Examples::
 .add_argument("indices", "NDArray-or-Symbol", "array of locations where to set on_value")
 .add_arguments(OneHotParam::__FIELDS__());
 
+NNVM_REGISTER_OP(sparse_retain)
+.describe(R"code(pick rows specified by user input index array from a row sparse matrix
```
Comment (eric-haibin-lin) [inline src/operator/tensor/indexing_op.cc]: Add an example here? Similar to the one in https://www.tensorflow.org/api_docs/python/tf/sparse_retain

File: tests/python/unittest/test_sparse_operator.py [hunk 2]
```diff
@@ -175,6 +179,32 @@ def check_csr_slice(shape, slice_input):
...
     check_csr_slice(shape, False)
 
+
+def test_sparse_retain():
+    for num_rows in range(1, 40):
+        num_cols = 3
+        shape = (num_rows, num_cols)
+        rsp, _ = rand_sparse_ndarray(shape=shape, storage_type='row_sparse', density=0.5)
+        length = np.random.randint(1, num_rows + 1)
+        import random
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_sparse_operator.py]: We seldom import `random`. You can use rnd.randint() instead, which is from numpy. There's also sampling logic in `def rand_sparse_ndarray`

File: tests/python/unittest/test_sparse_operator.py [hunk 3]
```diff
@@ -175,6 +179,32 @@ def check_csr_slice(shape, slice_input):
     check_csr_slice(shape, True)
     check_csr_slice(shape, False)
 
+
+def test_sparse_retain():
+    for num_rows in range(1, 40):
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_sparse_operator.py]: Not necessary to run 40 iteration in a single test. Consider rand_shape_2d()?

File: src/operator/tensor/indexing_op.cc (hunk 1 above)
Comment (reminisce) [inline src/operator/tensor/indexing_op.cc]: Done.

File: tests/python/unittest/test_sparse_operator.py (hunk 2 above)
Comment (reminisce) [inline tests/python/unittest/test_sparse_operator.py]: Added a random_sample function in test_utils.py.

File: tests/python/unittest/test_sparse_operator.py (hunk 3 above)
Comment (reminisce) [inline tests/python/unittest/test_sparse_operator.py]: Changed.

**Pull request merged**